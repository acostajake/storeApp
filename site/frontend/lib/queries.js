import gql from 'graphql-tag';

import { perPage } from '../config';

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$desc: String!
		$image: String
		$largeImg: String
		$price: Int!
	) {
		createItem(
			title: $title
			desc: $desc
			image: $image
			largeImg: $largeImg
			price: $price
		) {
			id
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$id: ID!
		$title: String
		$desc: String
		$price: Int
	) {
		updateItem(id: $id, title: $title, desc: $desc, price: $price) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ALL_ITEMS = gql`
	query GET_ALL_ITEMS($skip: Int = 0, $first: Int = ${perPage}) {
		items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
			id
			title
			desc
			image
			largeImg
			price
		}
	}
`;

export const GET_ITEM = gql`
	query GET_ITEM($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ITEM_WITH_IMAGE = gql`
	query GET_ITEM_WITH_IMAGE($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			largeImg
		}
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

// Auth
export const RESET_MUTATION = gql`
	mutation RESET_MUTATION(
		$resetToken: String!
		$password: String!
		$confirmPassword: String!
	) {
		resetPassword(
			resetToken: $resetToken
			password: $password
			confirmPassword: $confirmPassword
		) {
			id
			name
			email
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const SIGNUP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$name: String!
		$email: String!
		$password: String!
	) {
		signup(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNIN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signin(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNOUT_MUTATION = gql`
	mutation SIGNOUT_MUTATION {
		signout {
			message
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation updatePermissions($permissions: [Permission], $userId: ID!) {
		updatePermissions(permissions: $permissions, userId: $userId) {
			id
			permissions
			name
			email
		}
	}
`;

export const CURRENT_USER = gql`
	query {
		me {
			id
			email
			name
			permissions
		}
	}
`;

export const GET_ALL_USERS = gql`
	query {
		users {
			id
			name
			email
			permissions
		}
	}
`;
