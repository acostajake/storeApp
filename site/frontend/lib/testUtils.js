import casual from 'casual';
import {
	CURRENT_USER,
	PAGINATION_QUERY,
	REQUEST_RESET_MUTATION,
} from './queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'dogs are best',
	desc: 'dogs',
	largeImg: 'dog.jpg',
});

const fakeUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	desc: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: fakeItem(),
	user: fakeUser(),
	...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: null } },
	},
];

const signedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: fakeUser() } },
	},
];

const signedInWithCartMocks = [
	{
		request: { query: CURRENT_USER },
		result: {
			data: {
				me: {
					...fakeUser(),
					cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
				},
			},
		},
	},
];

const makeMocksFor = (length) => {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							__typename: 'count',
							count: length,
						},
					},
				},
			},
		},
	];
};

const requestResetMock = [
	{
		request: {
			query: REQUEST_RESET_MUTATION,
			variables: { email: 'test@fakeemail.com' },
		},
		result: {
			data: {
				requestReset: {
					message: 'success',
					__typename: 'Message',
				},
			},
		},
	},
];

export {
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
	notSignedInMocks,
	requestResetMock,
	signedInMocks,
	signedInWithCartMocks,
	makeMocksFor,
};
