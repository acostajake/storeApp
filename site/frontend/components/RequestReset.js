import React, { useState } from 'react';
import { Mutation } from 'react-apollo';

import Form from './styles/Form';
import Error from './ErrorMessage';

import { REQUEST_RESET_MUTATION } from '../lib/queries';

const RequestReset = () => {
	const [email, setEmail] = useState('');

	return (
		<Mutation mutation={REQUEST_RESET_MUTATION} variables={{ email }}>
			{(requestReset, { called, error, loading }) => {
				return (
					<Form
						data-test='form'
						method='POST'
						onSubmit={async (e) => {
							e.preventDefault();
							await requestReset();
							setEmail('');
						}}
					>
						<fieldset disabled={loading} aria-busy={loading}>
							<h2>Request password reset</h2>
							<Error error={error} />
							{!error && !loading && called && (
								<p>Success! Use the link sent to your email to reset.</p>
							)}
							<label htmlFor='email'>
								Email
								<input
									type='email'
									name='email'
									placeholder='myname@something.com'
									value={email}
									onChange={(e) => setEmail(e.target.value)}
								/>
							</label>
							<button type='submit'>Request reset</button>
						</fieldset>
					</Form>
				);
			}}
		</Mutation>
	);
};

export default RequestReset;
