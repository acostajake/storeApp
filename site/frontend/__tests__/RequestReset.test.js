import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import { requestResetMock } from '../lib/testUtils';
import RequestReset from '../components/RequestReset';

describe('<RequestReset/>', () => {
	it('renders and matches snapshot', async () => {
		const mountedReqReset = mount(
			<MockedProvider mocks={requestResetMock}>
				<RequestReset />
			</MockedProvider>
		);
		await wait();
		mountedReqReset.update();
		const form = mountedReqReset.find('[data-test="form"]');
		expect(toJSON(form)).toMatchSnapshot();
	});
	it('calls the mutation', async () => {
		const mountedReqReset = mount(
			<MockedProvider mocks={requestResetMock}>
				<RequestReset />
			</MockedProvider>
		);
		const success = 'Success! Use the link sent to your email to reset.';
		mountedReqReset.find('input').simulate('change', {
			target: { name: 'email', value: 'test@fakeemail.com' },
		});
		mountedReqReset.find('form').simulate('submit');
		await wait();
		mountedReqReset.update();
		expect(mountedReqReset.find('p').text()).toContain(success);
	});
});
