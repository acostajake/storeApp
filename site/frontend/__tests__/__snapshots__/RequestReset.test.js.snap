// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
Array [
  <Form
    data-test="form"
    method="POST"
    onSubmit={[Function]}
  >
    <form
      className="Form-opr7l5-0 gCHedR"
      data-test="form"
      method="POST"
      onSubmit={[Function]}
    >
      <fieldset
        aria-busy={false}
        disabled={false}
      >
        <h2>
          Request password reset
        </h2>
        <DisplayError
          error={Object {}}
        />
        <label
          htmlFor="email"
        >
          Email
          <input
            name="email"
            onChange={[Function]}
            placeholder="myname@something.com"
            type="email"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Request reset
        </button>
      </fieldset>
    </form>
  </Form>,
  <form
    className="Form-opr7l5-0 gCHedR"
    data-test="form"
    method="POST"
    onSubmit={[Function]}
  >
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <h2>
        Request password reset
      </h2>
      <DisplayError
        error={Object {}}
      />
      <label
        htmlFor="email"
      >
        Email
        <input
          name="email"
          onChange={[Function]}
          placeholder="myname@something.com"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request reset
      </button>
    </fieldset>
  </form>,
]
`;
